server/src/internalClusterTest/java/org/elasticsearch/get/ShardMultiGetFomTranslogActionIT.java

```diff
@@ -0,0 +1,149 @@
+        assertThat(getResponses.size(), equalTo(2));
+        assertNotNull("get followed by a delete should still return a result", getResponses.get(0));
+        assertThat(getResponses.get(0).getId(), equalTo("1"));
+        assertThat(getResponses.get(0).isExists(), equalTo(false));
+        assertNull(getResponses.get(1));
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+
+        indexResponse = client().prepareIndex("test").setSource("field1", "value2").get();
+        response = getFromTranslog(indexOrAlias(), indexResponse.getId());
+        multiGetShardResponse = response.multiGetShardResponse();
+        assertThat(getLocations(multiGetShardResponse).size(), equalTo(1));
+        assertTrue(getFailures(multiGetShardResponse).stream().allMatch(Objects::isNull));
+        getResponses = getResponses(multiGetShardResponse);
+        assertThat(getResponses.size(), equalTo(1));
+        assertNotNull(getResponses.get(0));
+        assertThat(getResponses.get(0).getId(), equalTo(indexResponse.getId()));
+        assertThat(getResponses.get(0).getVersion(), equalTo(indexResponse.getVersion()));
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+        // After a refresh we should not be able to get from translog
+        refresh("test");
+        response = getFromTranslog(indexOrAlias(), indexResponse.getId());
+        multiGetShardResponse = response.multiGetShardResponse();
+        assertThat(getLocations(multiGetShardResponse).size(), equalTo(1));
+        assertTrue(getFailures(multiGetShardResponse).stream().allMatch(Objects::isNull));
+        assertTrue(
+            "after a refresh we should not be able to get from translog",
+            getResponses(multiGetShardResponse).stream().allMatch(Objects::isNull)
+        );
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+        // After two refreshes the LiveVersionMap switches back to append-only and stops tracking IDs
+        // Refreshing with empty LiveVersionMap doesn't cause the switch, see {@link LiveVersionMap.Maps#shouldInheritSafeAccess()}.
+        client().prepareIndex("test").setSource("field1", "value3").get();
+        refresh("test");
+        refresh("test");
+        // An optimized index operation marks the maps as unsafe
+        client().prepareIndex("test").setSource("field1", "value4").get();
+        response = getFromTranslog(indexOrAlias(), "non-existent");
+        multiGetShardResponse = response.multiGetShardResponse();
+        assertThat(getLocations(multiGetShardResponse).size(), equalTo(1));
+        assertThat(getFailures(multiGetShardResponse).size(), equalTo(1));
+        assertNull(getFailures(multiGetShardResponse).get(0));
+        assertThat(getResponses(multiGetShardResponse).size(), equalTo(1));
+        assertNull(getResponses(multiGetShardResponse).get(0));
+        assertThat(response.segmentGeneration(), greaterThan(0L));
+    }
```

Comment:
So far I've seen only one id in this test being set and returned as a value. It would be nice to have at least one more id set that's actually returned by a multi get, and both values are asserted. For the moment, I see there's one case where you request 3 ids, but only the first one has a value, while the other 2 never get a value. So I think it would strengthen the test to have one more case where you set 2 ids and return them and assert them.

Reply:
https://github.com/elastic/elasticsearch/pull/96492/commits/40a3e325b169357c04c9786699bd8c7fac4c7f3d